mlflow/utils/file_utils.py

```diff
@@ -999,3 +1001,22 @@ def chdir(path: str) -> None:
         yield
     finally:
         os.chdir(cwd)
+
+
+def get_total_size(path: str) -> int:
+    """
+    Return the size of all files under given path, including files in subdirectories.
+
+    :param path: The absolute path of a local directory.
+    :return: size in bytes.
+    """
+    if not os.path.isdir(path):
+        raise MlflowException(
+            message=f"The given {path} is not a directory.", error_code=INVALID_PARAMETER_VALUE
+        )
+
+    total_size = 0
+    for cur_path, dirs, files in os.walk(path):
```

Comment:
Does this work for all file types? 

I tried this logic on my local MLflow filesystem storage. 

```python 
def get_total_size(path: str) -> int:
    """
    Return the size of all files under given path, including files in subdirectories.
    :param path: The absolute path of a local directory.
    :return: size in bytes.
    """
    if not os.path.isdir(path):
        raise MlflowException(
            message=f"The given {path} is not a directory.", error_code=INVALID_PARAMETER_VALUE
        )

    total_size = 0
    for cur_path, dirs, files in os.walk(path):
        full_paths = [os.path.join(cur_path, file) for file in files]
        total_size += sum([os.path.getsize(file) for file in full_paths])
    return total_size
```
(I have some large LLM's in here for local OSS testing)
``` python
get_total_size(os.path.expanduser("~/JupyterNotebooks"))
```

```bash
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
/var/folders/cd/n8n0rm2x53l_s0xv_j_xklb00000gp/T/ipykernel_37720/1521265667.py in <cell line: 1>()
----> 1 get_total_size(os.path.expanduser("~/JupyterNotebooks"))

/var/folders/cd/n8n0rm2x53l_s0xv_j_xklb00000gp/T/ipykernel_37720/3741259608.py in get_total_size(path)
     13     for cur_path, dirs, files in os.walk(path):
     14         full_paths = [os.path.join(cur_path, file) for file in files]
---> 15         total_size += sum([os.path.getsize(file) for file in full_paths])
     16     return total_size

/var/folders/cd/n8n0rm2x53l_s0xv_j_xklb00000gp/T/ipykernel_37720/3741259608.py in <listcomp>(.0)
     13     for cur_path, dirs, files in os.walk(path):
     14         full_paths = [os.path.join(cur_path, file) for file in files]
---> 15         total_size += sum([os.path.getsize(file) for file in full_paths])
     16     return total_size

~/miniconda3/envs/mlflow-dev-env/lib/python3.8/genericpath.py in getsize(filename)
     48 def getsize(filename):
     49     """Return the size of a file, reported by os.stat()."""
---> 50     return os.stat(filename).st_size
     51 
     52 

FileNotFoundError: [Errno 2] No such file or directory: '/Users/benjamin.wilson/JupyterNotebooks/mpt-7b/pytorch_model-00001-of-00002.bin'
```

Modifying the approach to use unix-native processing: 

```python
import subprocess
import os

def get_directory_size(path: str) -> int:
    if os.name == 'posix':  # Unix/Linux/macOS
        cmd = ['du', '-sk', path]
        output = subprocess.check_output(cmd)
        size = int(output.split()[0])
    elif os.name == 'nt':  # Windows
        cmd = ['robocopy', '/L', '/XJ', '/E', '/BYTES', '/NP', path, 'NULL']
        output = subprocess.check_output(cmd, universal_newlines=True)
        size_line = [line for line in output.splitlines() if "Bytes :" in line][0]
        size = int(size_line.split(":")[1].strip().replace(",", ""))
    else:
        raise ValueError("Unsupported OS")
    return size
```

``` python
get_directory_size(os.path.expanduser("~/JupyterNotebooks")) / 1000
```

```bash
27766.1
```

Reply:
Hmm weird - I tested on my local os and it worked
```
get_total_size(os.path.expanduser("~/transformers-cache/models--t5-large-resharded"))
```

I didn't choose  the option of using unix native commands because I'm not super clear what it's returning. It seems to return different director sizes when I do `-s` (which I thought returns bytes) v.s. `-sk` (which I thought returns kilobytes). See below...  
```
(mlflow-dev) wenfei.yan universe [master]$ du -s /Users/wenfei.yan/transformers-cache/models--t5-large-resharded
6277552	/Users/wenfei.yan/transformers-cache/models--t5-large-resharded
(mlflow-dev) wenfei.yan universe [master]$ du -sk /Users/wenfei.yan/transformers-cache/models--t5-large-resharded
3138776	/Users/wenfei.yan/transformers-cache/models--t5-large-resharded
```

Reply:
`-s` is for "summarize" (adds up the individual files sizes). 
-k would be kB, -m for MB, etc. 

Without putting a value in, it uses the default block size for the `du` operator (512bytes if POSIXLY_CORRECT is set as a sys env var, 1024 otherwise). 

I'm not recommending to use this; more very curious about why the implementation wasn't able to read a .bin file in my local filesystem for transformers models. We should definitely investigate thoroughly for correctness to ensure that size operations can work (and perform error handling if there are permissions errors or unreadable files for a given operating system)

Reply:
Yeah I'm also curious why it'd fail to read the .bin file in your case. Could you describe your setup? I'll try to reproduce.

It's actually a good call that we should try catch the error because failing to calculate the size should not fail the whole `save_model`.

Reply:
The model that failed to read was read using: 
``` bash
mlflow==2.7.1 
transformers==4.34.1
huggingface_hub==0.17.3
``` 

Python 3.8.10 (our min required dev environment for MLflow for compatibility, so that's where the `os` version comes from) 

OSX Ventura 13.5 
M1 Mac Pro


On the try catch front, what is your suggested approach if we get an incomplete read on the directory size? Do we omit the MLModel entry or provide an indicator that the file size could not be calculated properly? 
Is there a hard dependency that you're building on the consumption of the presence of this key downstream? Does typing matter for this field?

Reply:
My proposal is just omitting this entry in MLModel and optionally log a warning (maybe it's not even worth the "WARNING" level).
The downstream task should not have hard dependency. Anyway models logged from older mlflow versions won't have this model size in MLModel file.

Regarding for different file types and different platforms, my impression is that I can write some more test cases that cover different flavors (thus different file types) and CI will run it on various envs/platforms, and so the problem is covered. Is this correct?

Reply:
@BenWilson2 

Does `FileNotFoundError: [Errno 2] No such file or directory: '/Users/benjamin.wilson/JupyterNotebooks/mpt-7b/pytorch_model-00001-of-00002.bin'` exist?

Reply:
@harupy the file exists and I was able to load the saved transformers library with mlflow and with the transformers library directly by passing in the file path. du inspection works as well. I suspect that there are certain conditions where the python os operations don't always correctly inspect files on disk (this isn't the first time I've found this in my time using Python either).